File: Pascal's Triangle.py
# Leetcode Problem 118: Pascal's Triangle
# PYTHON CODE
class Solution:

    def generate(self, numRows: int) -> list[list[int]]:
        triangle = []

        for i in range(numRows):
            # First element of every row is always 1
            row = [1] * (i + 1)

            # Fill in the middle elements using the previous row
            for j in range(1, i):
                row[j] = triangle[i - 1][j - 1] + triangle[i - 1][j]

            triangle.append(row)

        return triangle

File: Pascal's Triangle.java
// Leetcode Problem 118: Pascal's Triangle
// JAVA CODE
import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<List<Integer>> generate(int numRows) {
        List<List<Integer>> triangle = new ArrayList<>();

        for (int i = 0; i < numRows; i++) {
            List<Integer> row = new ArrayList<>();

            for (int j = 0; j <= i; j++) {
                // First and last elements of each row are always 1
                if (j == 0 || j == i) {
                    row.add(1);
                } else {
                    // Sum the two elements directly above in the previous row
                    List<Integer> prevRow = triangle.get(i - 1);
                    int sum = prevRow.get(j - 1) + prevRow.get(j);
                    row.add(sum);
                }
            }

            triangle.add(row);
        }

        return triangle;
    }
}